examples/java/src/main/java/org/apache/beam/examples/WindowedWordCount.java

```diff
@@ -192,44 +199,42 @@ public static void main(String[] args) throws IOException {
      * windows over a PCollection containing windowed values.
      */
-    PCollection<KV<String, Long>> wordCounts = windowedWords.apply(new WordCount.CountWords());
-
-    /**
-     * Concept #5: Customize the output format using windowing information
-     *
-     * <p>At this point, the data is organized by window. We're writing text files and and have no
-     * late data, so for simplicity we can use the window as the key and {@link GroupByKey} to get
-     * one output file per window. (if we had late data this key would not be unique)
-     *
-     * <p>To access the window in a {@link DoFn}, add a {@link BoundedWindow} parameter. This will
-     * be automatically detected and populated with the window for the current element.
-     */
-    PCollection<KV<IntervalWindow, KV<String, Long>>> keyedByWindow =
-        wordCounts.apply(
-            ParDo.of(
-                new DoFn<KV<String, Long>, KV<IntervalWindow, KV<String, Long>>>() {
-                  @ProcessElement
-                  public void processElement(ProcessContext context, IntervalWindow window) {
-                    context.output(KV.of(window, context.element()));
-                  }
-                }));
-
-    /**
-     * Concept #6: Format the results and write to a sharded file partitioned by window, using a
-     * simple ParDo operation. Because there may be failures followed by retries, the
-     * writes must be idempotent, but the details of writing to files is elided here.
-     */
-    keyedByWindow
-        .apply(GroupByKey.<IntervalWindow, KV<String, Long>>create())
-        .apply(ParDo.of(new WriteWindowedFilesDoFn(output)));
+    PCollection<KV<String, Long>> wordCounts = windowedWords.apply(
+        "CountWords", new WordCount.CountWords());
+
+    /** Concept #5: Format the results and write to a sharded file partitioned by window. */
+    wordCounts
+        .apply("FormatOutput", MapElements.via(new FormatAsTextFn()))
+        .apply(
+            "WriteOutput",
+            TextIO.Write.to(
+                    new FilenamePolicy() {
+                      @Override
+                      public ValueProvider<String> getBaseOutputFilenameProvider() {
+                        return StaticValueProvider.of(output);
+                      }
+
+                      @Override
+                      public String apply(Context input) {
+                        IntervalWindow window = (IntervalWindow) input.getWindow();
+                        return String.format(
+                            "%s-%s-%s",
+                            output, formatter.print(window.start()), formatter.print(window.end()));
+                      }
+                    })
+                // This is the number of shards per window. We set it to 1 here for testing
+                // convenience, however for many runners this does not need to be set. The output
+                // is still sharded by window however.
+                .withNumShards(1)
```

Comment:
This will have problematic performance on runners that only parallelize over keys, which is all runners. Previously the windows were in reified the key so there would be parallelism available.